// stylelint/dist/rules/property-no-unknown/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messages = exports.ruleName = undefined;

exports.default = function (actual, options) {
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, { actual: actual }, {
      actual: options,
      possible: {
        ignoreProperties: [_lodash.isString]
      },
      optional: true
    });

    if (!validOptions) {
      return;
    }

    root.walkDecls(function (decl) {
      var prop = decl.prop;


      if (!(0, _utils.isStandardSyntaxProperty)(prop)) {
        return;
      }
      if ((0, _utils.isCustomProperty)(prop)) {
        return;
      }

      if ((0, _utils.optionsHaveIgnoredProperty)(options, prop)) {
        return;
      }

      if (_knownCssProperties.all.indexOf(prop.toLowerCase()) !== -1) {
        return;
      }

      (0, _utils.report)({
        message: messages.rejected(prop),
        node: decl,
        result: result,
        ruleName: ruleName
      });
    });
  };
};

var _utils = require("../../utils");

var _lodash = require("lodash");

var _knownCssProperties = require("known-css-properties");

var ruleName = exports.ruleName = "property-no-unknown";

var messages = exports.messages = (0, _utils.ruleMessages)(ruleName, {
  rejected: function rejected(property) {
    return "Unexpected unknown property \"" + property + "\"";
  }
});
